### PR TITLE
Improved key columns in aws_cloudwatch_log_stream table and added the columns log_stream_name_prefix, descending, and order_by Closes #1949

### LINKED ISSUE #1949: Improve key columns in `aws_cloudwatch_log_stream` table
**Is your feature request related to a problem? Please describe.**
When querying a large amount of log streams, the API limits cause a lot of throttling

**Describe the solution you'd like**
Better quals based on what's available for the API action, https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_DescribeLogStreams.html

**Describe alternatives you've considered**
Using Steampipe rate limiters or increasing the number of retries

**Additional context**

### PR DESCRIPTION
**Pull request by ParthaI**
# Integration test logs
<details>
  <summary>Logs</summary>

```
N/A
```
</details>

# Example query results
<details>
  <summary>Results</summary>

```
> select name, log_group_name, log_stream_name_prefix, descending, order_by from aws_cloudwatch_log_stream;
+------------------------------------------------------------------+-----------------------------------------------------------------------+------------------------+------------+---------------+
| name                                                             | log_group_name                                                        | log_stream_name_prefix | descending | order_by      |
+------------------------------------------------------------------+-----------------------------------------------------------------------+------------------------+------------+---------------+
| 2023-06-22#0E975B7024ACC6F69CB81D4C11A51DE6                      | /aws/elasticbeanstalk/Test53-env/environment-health.log               | <null>                 | false      | LogStreamName |
| 2022/12/01/[$LATEST]f0e3f1c779f24b718eeaacea1484d8ae             | /aws/lambda/SecurityLake_Glue_Partition_Updater_Lambda_ap-northeast-1 | <null>                 | false      | LogStreamName |
| 2022/12/01/[$LATEST]2603f14ee1b64df496e54c3b7e6df1e3             | /aws/lambda/SecurityLake_Glue_Partition_Updater_Lambda_ap-northeast-1 | <null>                 | false      | LogStreamName |

> select name, log_group_name, log_stream_name_prefix, descending, order_by from aws_cloudwatch_log_stream where name = 'BackupDelivery'
+----------------+-----------------------------------+------------------------+------------+---------------+
| name           | log_group_name                    | log_stream_name_prefix | descending | order_by      |
+----------------+-----------------------------------+------------------------+------------+---------------+
| BackupDelivery | /aws/kinesisfirehose/PUT-S3-FBs5h | <null>                 | false      | LogStreamName |
+----------------+-----------------------------------+------------------------+------------+---------------+

> select name, log_group_name, log_stream_name_prefix, descending, order_by from aws_cloudwatch_log_stream where name = 'BackupDelivery' and log_stream_name_prefix = 'aws-neptune'
+------+----------------+------------------------+------------+----------+
| name | log_group_name | log_stream_name_prefix | descending | order_by |
+------+----------------+------------------------+------------+----------+
+------+----------------+------------------------+------------+----------+

Time: 8.2s. Rows fetched: 2. Hydrate calls: 0.
```
</details>

### CODE REVIEW

**Comment by ParthaI** [comment]
This PR should work per the suggestion provided in this [issue comment](https://github.com/turbot/steampipe-plugin-aws/issues/1949#issuecomment-1779719366).

**Case 1:** If a user specifies `name = '2023-06-22#0E975B7024ACC6F69CB81D4C1151DE6'`, use it as LogStreamNamePrefix(**Result: Successful API call**)
```
>  select * from aws_cloudwatch_log_stream where name = '2023-06-22#0E975B7024ACC6F69CB81D4C11A51DE6'
+---------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------------------+---->
| name                                        | arn                                                                                                                                                           | log_group_name                                          | log>
+---------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------------------+---->
| 2023-06-22#0E975B7024ACC6F69CB81D4C11A51DE6 | arn:aws:logs:ap-south-1:xxxxxxxxxxxx:log-group:/aws/elasticbeanstalk/Test53-env/environment-health.log:log-stream:2023-06-22#0E975B7024ACC6F69CB81D4C11A51DE6 | /aws/elasticbeanstalk/Test53-env/environment-health.log | <nu>
+---------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------------------+---->

Time: 3.4s. Rows fetched: 1. Hydrate calls: 1.
```

**Case 2:** If a user specifies `log_stream_name_prefix = '2023-06-22'`, use it as `LogStreamNamePrefix`(**Result: Successful API call**)
```
> select * from aws_cloudwatch_log_stream where log_stream_name_prefix = '2023-06-22'
+---------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------------------------------------------>
| name                                        | arn                                                                                                                                                             | log_group_name                                             >
+---------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------------------------------------------>
| 2023-06-22#0E975B7024ACC6F69CB81D4C11A51DE6 | arn:aws:logs:ap-south-1:xxxxxxxxxxxx:log-group:/aws/elasticbeanstalk/Test53-env/environment-health.log:log-stream:2023-06-22#0E975B7024ACC6F69CB81D4C11A51DE6   | /aws/elasticbeanstalk/Test53-env/environment-health.log    >
| 2023-06-22#B5A100C12FEE498324A234B9ED377644 | arn:aws:logs:us-east-1:xxxxxxxxxxxx:log-group:/aws/elasticbeanstalk/Test32-envtwe/environment-health.log:log-stream:2023-06-22#B5A100C12FEE498324A234B9ED377644 | /aws/elasticbeanstalk/Test32-envtwe/environment-health.log >
+---------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------------------------------------------------>

Time: 3.3s. Rows fetched: 2. Hydrate calls: 2.
```

**Case 3:** If a user specifies` log_stream_name_prefix = '2023-06-22' and name = '2023-06-22#0E975B7024ACC6F69CB81D4C1151DE6'`, use `log_stream_name_prefix` as LogStreamNamePrefix(**Result: Successful API call**)
```
> select * from aws_cloudwatch_log_stream where log_stream_name_prefix = '2023-06-22' and name = '2023-06-22#0E975B7024ACC6F69CB81D4C11A51DE6';
+---------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------------------+---->
| name                                        | arn                                                                                                                                                           | log_group_name                                          | log>
+---------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------------------+---->
| 2023-06-22#0E975B7024ACC6F69CB81D4C11A51DE6 | arn:aws:logs:ap-south-1:xxxxxxxxxxxx:log-group:/aws/elasticbeanstalk/Test53-env/environment-health.log:log-stream:2023-06-22#0E975B7024ACC6F69CB81D4C11A51DE6 | /aws/elasticbeanstalk/Test53-env/environment-health.log | 202>
+---------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------------------+---->

Time: 3.3s. Rows fetched: 2. Hydrate calls: 2.
```

**Case 4:** If a user specifies `order_by = 'LastEventTime'`, use it as OrderBy(**Result: Successful API call**)
```
> select * from aws_cloudwatch_log_stream where order_by = 'LastEventTime'
+------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------->
| name                                                             | arn                                                                                                                                                                                      | log_group_nam>
+------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------->
| 2023-06-22#0E975B7024ACC6F69CB81D4C11A51DE6                      | arn:aws:logs:ap-south-1:xxxxxxxxxxxx:log-group:/aws/elasticbeanstalk/Test53-env/environment-health.log:log-stream:2023-06-22#0E975B7024ACC6F69CB81D4C11A51DE6                            | /aws/elasticb>
| 2022/12/01/[$LATEST]20e9d259063e42d7a44561c0314daa64             | arn:aws:logs:ap-northeast-1:xxxxxxxxxxxx:log-group:/aws/lambda/SecurityLake_Glue_Partition_Updater_Lambda_ap-northeast-1:log-stream:2022/12/01/[$LATEST]20e9d259063e42d7a44561c0314daa64 | /aws/lambda/S>
| 2022/12/01/[$LATEST]7ccb7eeae8734498b6f04c978fa70a40             | arn:aws:logs:ap-northeast-1:xxxxxxxxxxxx:log-group:/aws/lambda/SecurityLake_Glue_Partition_Updater_Lambda_ap-northeast-1:log-stream:2022/12/01/[$LATEST]7ccb7eeae8734498b6f04c978fa70a40 | /aws/lambda/S>
| 2022/12/01/[$LATEST]c94b51b0b46645b1ba3c7de18d4d5aa5             | arn:aws:logs:ap-northeast-1:xxxxxxxxxxxx:log-group:/aws/lambda/SecurityLake_Glue_Partition_Updater_Lambda_ap-northeast-1:log-stream:2022/12/01/[$LATEST]c94b51b0b46645b1ba3c7de18d4d5aa5 | /aws/lambda/S>

Time: 3.3s. Rows fetched: 166. Hydrate calls: 166.
```

**Case 5:** If a user specifies `name = '2023-06-22#0E975B7024ACC6F69CB81D4C1151DE6' and order_by = 'LastEventTime'`, use `OrderBy` but not `LogStreamNamePrefix`(**Result: Successful API call**)
```
> select * from aws_cloudwatch_log_stream where  name = '2023-06-22#0E975B7024ACC6F69CB81D4C11A51DE6' and order_by = 'LastEventTime';
+---------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------------------+---->
| name                                        | arn                                                                                                                                                           | log_group_name                                          | log>
+---------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------------------+---->
| 2023-06-22#0E975B7024ACC6F69CB81D4C11A51DE6 | arn:aws:logs:ap-south-1:xxxxxxxxxxxx:log-group:/aws/elasticbeanstalk/Test53-env/environment-health.log:log-stream:2023-06-22#0E975B7024ACC6F69CB81D4C11A51DE6 | /aws/elasticbeanstalk/Test53-env/environment-health.log | <nu>
+---------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------------------------------+---->

Time: 3.4s. Rows fetched: 166. Hydrate calls: 166.
```

**Case 6:** If a user specifies `log_stream_name_prefix = '2023-06-22' and order_by = 'LastEventTime',` use `OrderBy` and `LogStreamNamePrefix`(**Result: Error from SDK**)
```
> select * from aws_cloudwatch_log_stream where log_stream_name_prefix = '2023-06-22' and order_by = 'LastEventTime'

Error: operation error CloudWatch Logs: DescribeLogStreams, https response error StatusCode: 400, RequestID: e2dc9afd-166a-4776-a84f-a8623e428de5, InvalidParameterException: Cannot order by LastEventTime with a logStreamNamePrefix. (SQLSTATE HV000)

+------+-----+----------------+------------------------+------------+----------+---------------+-----------------------+----------------------+---------------------+-----------------------+-------+------+-----------+--------+------------+------+
| name | arn | log_group_name | log_stream_name_prefix | descending | order_by | creation_time | first_event_timestamp | last_event_timestamp | last_ingestion_time | upload_sequence_token | title | akas | partition | region | account_id | _ctx |
+------+-----+----------------+------------------------+------------+----------+---------------+-----------------------+----------------------+---------------------+-----------------------+-------+------+-----------+--------+------------+------+
+------+-----+----------------+------------------------+------------+----------+---------------+-----------------------+----------------------+---------------------+-----------------------+-------+------+-----------+--------+------------+------+

Time: 1.0s.
```

**Case 7:** If a user specifies `log_stream_name_prefix = '2023-06-22' and name = '2023-06-22#0E975B7024ACC6F69CB81D4C1151DE6' and order_by = 'LastEventTime'`, use `OrderBy` and  `log_stream_name_prefix` as `LogStreamNamePrefix`(**Result: Error from SDK**)
```
> select * from aws_cloudwatch_log_stream where log_stream_name_prefix = '2023-06-22' and name = '2023-06-22#0E975B7024ACC6F69CB81D4C1151DE6' and order_by = 'LastEventTime'

Error: operation error CloudWatch Logs: DescribeLogStreams, https response error StatusCode: 400, RequestID: fd88f5f2-a38b-44aa-aabf-104aac1088c0, InvalidParameterException: Cannot order by LastEventTime with a logStreamNamePrefix. (SQLSTATE HV000)

+------+-----+----------------+------------------------+------------+----------+---------------+-----------------------+----------------------+---------------------+-----------------------+-------+------+-----------+--------+------------+------+
| name | arn | log_group_name | log_stream_name_prefix | descending | order_by | creation_time | first_event_timestamp | last_event_timestamp | last_ingestion_time | upload_sequence_token | title | akas | partition | region | account_id | _ctx |
+------+-----+----------------+------------------------+------------+----------+---------------+-----------------------+----------------------+---------------------+-----------------------+-------+------+-----------+--------+------------+------+
+------+-----+----------------+------------------------+------------+----------+---------------+-----------------------+----------------------+---------------------+-----------------------+-------+------+-----------+--------+------------+------+

Time: 1.4s.
```

**Review by madhushreeray30 — COMMENTED** [review]
@ParthaI please take a look at the comment, thanks!

**Review by madhushreeray30 — APPROVED** [review]

**Comment by misraved** [comment]
@ParthaI is there a reason for not including `logGroupIdentifier` as an optional qual?

Reference doc - https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_DescribeLogStreams.html#API_DescribeLogStreams_RequestSyntax

File: aws/table_aws_cloudwatch_log_stream.go [hunk 1]
```diff
@@ -146,8 +178,32 @@ func listCloudwatchLogStreams(ctx context.Context, d *plugin.QueryData, h *plugi
...
+	// Assign LogStreamNamePrefix if order is not by LastEventTime.
+	if !isOrderedByEventTime {
+		if name, exists := equalQuals["name"]; exists {
+			input.LogStreamNamePrefix = aws.String(name.GetStringValue())
+		} else if prefix, exists := equalQuals["log_stream_name_prefix"]; exists {
+			input.LogStreamNamePrefix = aws.String(prefix.GetStringValue())
+		}
+	} else if _, exists := equalQuals["log_stream_name_prefix"]; exists {
+		// If ordered by LastEventTime and log_stream_name_prefix is specified, use it.
+		input.LogStreamNamePrefix = aws.String(equalQuals["log_stream_name_prefix"].GetStringValue())
```
[the remark this answers is not in the document]
Comment (ParthaI) [reply aws/table_aws_cloudwatch_log_stream.go]: @madhushreeray30, indeed, `EqualsQualString` is a viable option. However, to maintain consistency across the table codes, especially for other column types like Boolean, I opted for using `equalQuals` instead.

**Comment by misraved** [comment]
@ParthaI Suggestions for the PR:
- Add `logGroupIdentifier` as an optional qual. We need to add `log_group_identifier` as a column as well. Valiadaton of this column should be done against the log group ARN.
- If a user passes in both `log_group_name` and `log_group_identifier` return an error per https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_DescribeLogStreams.html#API_DescribeLogStreams_RequestSyntax

**Comment by misraved** [comment]
@ParthaI as per discussion, since we do not get the column in the `GetConfig`, it is better to avoid the addition of the `log_goup_identifier` column to the table and as an optional qual.

**Pull request merged**